src/Compilers/Core/Portable/Operations/Operation.cs

```diff
@@ -138,22 +129,22 @@ public string Language
 
         protected void SetParentOperation(IOperation? parent)
         {
+            Debug.Assert(Volatile.Read(ref _parentDoNotAccessDirectly) == s_unset);
```

Comment:
>Debug.Assert(Volatile.Read(ref _parentDoNotAccessDirectly) == s_unset); [](start = 12, length = 71)

The fact that we are using ```Interlocked.CompareExchange``` below leads me to believe that a concurrency is possible. Given that, it feels like this assert can fail in some scenarios. If this method can no longer be called concurrently, we probably should adjust its implementation accordingly including the ```Volatile.Read``` usage.